### PR TITLE
Feature/request password reset

### PR DESCRIPTION
**Pull request by mwnicks**
### What

Added endpoint to handle requests for password reset emails 

### How to review

Ensure the unit and feature tests are passing.

*With a user created and confirmed within Cognito
- Send a POST request to '/v1/password-reset' - the body of the request should contain:
    - 'email': the email of the user
- You should receive a 202 response and an email with a verification code

- Send a POST request to '/v1/password-reset' - the body of the request should contain:
    - 'email': an email that does not relate to a user in Cognito
- You should receive a 202 response but no email

### Who can review

Anyone but me

### CODE REVIEW

File: api/users.go [hunk 1]
```diff
@@ -127,3 +127,44 @@ func (api *API) ChangePasswordHandler(ctx context.Context, w http.ResponseWriter
 
 	return models.NewSuccessResponse(jsonResponse, http.StatusAccepted, headers), nil
 }
+
+//PasswordResetHandler creates a new user and returns a http handler interface
```
Comment (sandypadmanabhan) [inline api/users.go]: comment to be updated

File: api/users.go [hunk 2]
```diff
@@ -127,3 +127,44 @@ func (api *API) ChangePasswordHandler(ctx context.Context, w http.ResponseWriter
...
+	if validationErr != nil {
+		return nil, models.NewErrorResponse([]error{validationErr}, http.StatusBadRequest, nil)
+	}
+
+	forgotPasswordRequest := passwordResetParams.BuildCognitoRequest(api.ClientSecret, api.ClientId)
+
+	result, err := api.CognitoClient.ForgotPassword(forgotPasswordRequest)
+	if err != nil {
+		responseErr := models.NewCognitoError(ctx, err, "ForgotPassword request from password reset endpoint")
+		if responseErr.Code == models.InternalError {
```
Comment (sandypadmanabhan) [inline api/users.go]: cognito response has 2 error code 400 and 500, but we are checking for just one here

File: api/users_test.go [hunk 3]
```diff
@@ -378,3 +379,115 @@ func TestChangePasswordHandler(t *testing.T) {
...
+					awsOrigErr := errors.New(awsErrCode)
+					awsErr := awserr.New(awsErrCode, awsErrMessage, awsOrigErr)
+					return nil, awsErr
+				},
+				http.StatusInternalServerError,
+			},
+			{
+				// Cognito invalid user
+				func(input *cognitoidentityprovider.ForgotPasswordInput) (*cognitoidentityprovider.ForgotPasswordOutput, error) {
+					awsErrCode := "CodeMismatchException"
```
Comment (sandypadmanabhan) [inline api/users_test.go]: https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ForgotPassword.html- Cannot find "CodeMismatchException". Am I looking at the correct list

File: api/users.go (hunk 1 above)
Comment (mwnicks) [reply api/users.go]: Updated to correctly describe the function

File: api/users_test.go (hunk 3 above)
Comment (mwnicks) [reply api/users_test.go]: I change the test scenario but missed changing the AWS Error code.
Should be 'UserNotFoundException', I've updated it now.
Tests were passing as they are both 400 errors handled in the same way in the handler.

**Review by sandypadmanabhan — COMMENTED** [review]
Do we want to update the swagger as well?

File: api/users.go [hunk 4]
```diff
@@ -127,3 +127,39 @@ func (api *API) ChangePasswordHandler(ctx context.Context, w http.ResponseWriter
...
 	return models.NewSuccessResponse(jsonResponse, http.StatusAccepted, headers), nil
 }
+
+//PasswordResetHandler requests a password reset email be sent to the user and returns a http handler interface
+func (api *API) PasswordResetHandler(ctx context.Context, w http.ResponseWriter, req *http.Request) (*models.SuccessResponse, *models.ErrorResponse) {
+	defer req.Body.Close()
+
+	body, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, handleBodyReadError(ctx, err)
```
Comment (sandypadmanabhan) [inline api/users.go:137]: We might need to log the event when an error is returned unless we have decided not to...

Comment (mwnicks) [reply api/users.go:137]: The new error constructors write a log entry for the error when the error is built.
See here https://github.com/ONSdigital/dp-identity-api/blob/develop/models/errors.go#L23

File: swagger.yaml [hunk 5]
```diff
@@ -223,6 +223,38 @@ paths:
...
+        202:
+          description: "Request accepted"
+        400:
+          description: "Email was not in a valid format"
+          schema:
+            $ref: '#/definitions/ErrorResponse'
+        500:
+          description: "Invalid request body or unexpected Cognito response"
+          schema:
+            $ref: '#/definitions/ErrorResponse'
```
Comment (sandypadmanabhan) [inline swagger.yaml:258]: can we add usernotfound error to the error object list

Comment (mwnicks) [reply swagger.yaml:258]: Good spot, I've added it now

**Review by sandypadmanabhan — APPROVED** [review]
LGTM

**Pull request merged**